add support for external ssh signing

seanaye opened the pull request:
<!---
Thank you for contributing to GitUI! Please fill out the template below, and remove or add any
information as you feel necessary.
--->

This Pull Request fixes/closes #2188.

It changes the following:
- Adds support for external bin signing with ssh keys


EDIT: I have added a test for the new feature, let me know if this is not sufficient or what to change, its a bit hard to test external binaries.

I ran make check without errors _except_ for the python check, I don't have python on my system which I'm guessing is the cause of the failure

I followed the checklist:
- [x] I added unittests
- [x] I ran `make check` without errors
- [x] I tested the overall application
- [x] I added an appropriate item to the changelog

seanaye committed:
add support for external ssh signing

extrawurst commented:
@yanganto you are the most prominent user of ssh-signing. can you review this?

extrawurst commented:
@seanaye is there a way to write a test for this?

seanaye commented:
I'll try to write some tests this week

kucho commented:
Very excited about this feature to start using gitui as my daily driver! Thank you, @seanaye 🙏

b1nhack commented:
Thanks a lot, I'm waiting for this feature.

dkarter commented:
Thank you for working on this! would love to see it released

seanaye commented:
@robrecord no worries! I actually totally forgot about this, I think I can find some time this week to finish it up

seanaye committed:
add test

seanaye committed:
update test

seanaye committed:
update changelog

seanaye commented:
@extrawurst @yanganto please let me know what else I can do to help get this merged

yanganto commented:
> @extrawurst @yanganto please let me know what else I can do to help get this merged

Hi @seanaye, 
The previous suggestion about no additional struct in the PR may not help you merge this PR, but it is good to respond if you don't think we need any changes.

extrawurst commented:
@kucho @dkarter @robrecord anyone who can help by building this branch and testing it with their setup and reporting back here can help get this merged.

seanaye commented:
> > @extrawurst @yanganto please let me know what else I can do to help get this merged
> 
> 
> 
> Hi @seanaye, 
> 
> The previous suggestion about no additional struct in the PR may not help you merge this PR, but it is good to respond even if you don't think we need any changes. 

I'm not quite sure what you mean by this, I don't see any previous feedback. If you review the PR I'm happy to make changes

yanganto commented on asyncgit/src/sync/sign.rs:
```diff
@@ -278,6 +324,144 @@ pub struct SSHSign {
 	secret_key: PrivateKey,
 }
```
Sorry, I found out the review is in stash not publish.

seanaye replied:
@yanganto Thanks for your comments. In my opinion having 2 smaller, but higher quality structs is preferable over having a single large struct. With this new suggestion the type system can encode invalid information for example 

```
program: None,
tmp_file: None,
secret_key: None
```

would be a representable type but is not valid. 

I also find that having more, smaller structs has better separation of concerns because each struct has a smaller distinct meaning.

If you feel strongly about your suggestions I can make the change but in my opinion this might make the code harder to maintain, not easier.

yanganto replied:
It is a suggestion, I did not againat this PR if you think the another struct is good.  It also works for me.

seanaye replied:
I will do which ever one you prefer as you are a maintainer. I wrote code that I would prefer to maintain but it is your choice.

extrawurst replied:
An extra struct seems reasonable to me if it bundles up separated logic. If all other concerns are considered please resolve this thread @yanganto

yanganto commented:
Does it work with `gpg.program = gpg2`?
What program have we already tested?  

It will be good to have a CI if we want to make sure this feature always works in the future.

seanaye commented:
> Does it work with `gpg.program = gpg2`?
> 
> What program have we already tested?  
> 
> 
> 
> It will be good to have a CI if we want to make sure this feature always works in the future.

I have tested so far with the 1Password ssh agent but the CLI arguments are standardized as far as I know

https://developer.1password.com/docs/ssh/git-commit-signing/

I think adding this to CI may be difficult as it would require adding new system dependencies to the image where the CI runs. For example installing the 1Password CLI inside a docker image.

yonas commented:
Ping

yanganto commented:
I thought the `dbg!`, `eprintln!`, and `debug_cmd_print!` in this project is only in used in test cases, if I am wrong please also let me know.  Thanks.  :pray:

seanaye committed:
remove dbg

seanaye commented:
Sorry for the late reply, I have removed the dbg and rebased to fix conflicts. I did have this working locally at one point but I have changed my workflow to no longer use `git` and thus I can no longer use gitui. 

The issue that you encountered would suggest that the program is not correctly reading the gpg format ssh. I would maybe suggest trying to insert a debug statement and rebuild to see what the value is at runtime. 

This is unfortunately the furthest I can contribute to this PR due to other priorities coming up. Please feel free to either continue this work or close

seanaye commented:
If you are curious about life after git, I am using [jj](https://github.com/jj-vcs/jj) on both personal and professional projects
